src/recoil/dotplot.ts

```diff
@@ -0,0 +1,197 @@
+  key: "diffexpGenesDotplot",
+  get: ({ get }) => {
+    const rows = get(dotplotRowState);
+    const ontology = get(currentOntologyState);
+
+    if (rows && ontology) {
+      const allDiffexpGenes = rows.map((row) => {
+        return ontology?.get(row.id)?.genes;
+      });
+
+      // flatten the array of arrays, remove duplicates and undefined values
+      const diffexpGenes = Array.from(new Set(allDiffexpGenes.flat().filter((genes) => genes !== undefined)));
+
+      return diffexpGenes;
+    } else {
+      return [];
+    }
+  },
+});
+
+// all of the celltypes as an array of strings like CL:0000000
+export const includedCelltypesState = selector<string[]>({
+  key: "includedCelltypes",
+  get: ({ get }) => {
+    const dotplotRows = get(dotplotRowState);
+
+    if (dotplotRows) {
+      return dotplotRows.map((row) => row.id);
+    } else {
+      return [];
+    }
+  },
+});
+
+/* 
+
+    each gene is a pair of cell/gene that looks like this
+    [
+      0 "10100", index position
+      1 "CL:0000065", cl term
+      2 "ENSMUSG00000045658", gene id
+      3 "2.2974446", mean
+      4 "0.59615386", frac expressing
+    ]
+
+    mean: geneData[i][3],
+    frac: geneData[i][4],
+  
+   */
+
+export const includedPairsState = selector<string[][]>({
+  key: "includedPairs",
+  get: ({ get }) => {
+    const geneData = get(geneDataState);
+    const includedCelltypes = get(includedCelltypesState);
+    const diffexpGenes = get(diffexpGenesDotplotState);
+
+    const includedPairs = geneData.filter((gene: string[]) => {
+      return includedCelltypes.includes(gene[1]) && diffexpGenes.includes(gene[2]);
+    });
+
+    return includedPairs;
+  },
+});
+
+interface DotplotData {
+  [key: string]: {
+    [key: string]: {
+      fracExpressing: string;
+      mean: string;
+    };
+  };
+}
+
+// the cell / gene pairs data in the dotplot, expression values
+export const dotplotState = selector<DotplotData>({
+  key: "includedGenes",
+  get: ({ get }) => {
+    const includedPairs = get(includedPairsState);
+    // return the pairs in the format { 'ENSG00000196549': { 'CL:0000000': { 'fracExpressing': 0.5, mean: 0.5 } } }
+    const pairsAsObject: DotplotData = {};
+
+    includedPairs.forEach((pair: string[]) => {
+      const gene = pair[2];
+      const celltype = pair[1];
+      const fracExpressing = pair[4];
+      const mean = pair[3];
+
+      if (!pairsAsObject[gene]) {
+        pairsAsObject[gene] = {};
+      }
+
+      pairsAsObject[gene][celltype] = {
+        fracExpressing,
+        mean,
+      };
+    });
+
+    return pairsAsObject;
+  },
+});
+
+export const dotScaleState = selector({
+  key: "dotColorScale",
+  get: ({ get }) => {
+    const includedPairs = get(includedPairsState);
+    const dotSizeMinRadius = get(dotSizeMinRadiusState);
+    const dotSizeMaxRadius = get(dotSizeMaxRadiusState);
+
+    const maxMean = Math.max(...includedPairs.map((pair) => parseFloat(pair[3])));
```

Comment:
I think it's okay for now, but it might be better to scale max and min per column. And the legend would be replaced with "min --- max" since we wouldn't have a global min and max anymore.